### PR TITLE
ARROW-215: Support other integer types in Parquet I/O

### PR DESCRIPTION
**Pull request by xhochy**
Code Review on Gerrit: https://gerrit.cloudera.org/#/c/3481/

### CODE REVIEW

File: cpp/src/arrow/parquet/parquet-io-test.cc [hunk 1]
```diff
@@ -52,25 +53,104 @@ template <typename TestType>
...
+  static constexpr LogicalType::type logical_enum = LogicalType::NONE;
+  static uint8_t const value;
+};
+
+const uint8_t test_traits<BooleanType>::value(1);
+
+template <>
+struct test_traits<UInt8Type> {
+  static constexpr ParquetType::type parquet_enum = ParquetType::INT32;
+  static constexpr LogicalType::type logical_enum = LogicalType::UINT_8;
```
Comment (wesm) [inline cpp/src/arrow/parquet/parquet-io-test.cc]: We should look at whether using the Parquet 2.0 logical types will cause problems in other readers. e.g. Impala doesn't have these ConvertedType enums

https://github.com/cloudera/Impala/blob/cdh5-trunk/common/thrift/parquet.thrift

@julienledem can you advise us here?

What other systems do is deal with converting Parquet's physical INT32 to smaller types through a layering of additional metadata -- for example, if a Hive or Impala table has `SMALLINT` or `TINYINT` type, then the int32 values are materilized into int16 or int8 slots

Comment (wesm) [inline cpp/src/arrow/parquet/parquet-io-test.cc]: What we might need to do is always return INT32 but provide an API where the user can request a particular column in the file to be casted to some other integer type.

File: cpp/src/arrow/parquet/writer.cc [hunk 2]
```diff
@@ -70,36 +72,124 @@ Status FileWriter::Impl::NewRowGroup(int64_t chunk_size) {
...
+    return Status::OK();                                          \
+  }
+
+CONVERT_PHYSICAL_TYPE_COPY_DATA(uint8_t, int32_t)
+CONVERT_PHYSICAL_TYPE_COPY_DATA(int8_t, int32_t)
+CONVERT_PHYSICAL_TYPE_COPY_DATA(uint16_t, int32_t)
+CONVERT_PHYSICAL_TYPE_COPY_DATA(int16_t, int32_t)
+CONVERT_PHYSICAL_TYPE_COPY_PTR(uint32_t, int32_t)
+CONVERT_PHYSICAL_TYPE_COPY_PTR(int32_t, int32_t)
+CONVERT_PHYSICAL_TYPE_COPY_PTR(uint64_t, int64_t)
```
Comment (wesm) [inline cpp/src/arrow/parquet/writer.cc]: Is this right for the unsigned integers?

File: cpp/src/arrow/parquet/reader.cc [hunk 3]
```diff
@@ -155,11 +159,53 @@ FlatColumnReader::Impl::Impl(MemoryPool* pool, const ::parquet::ColumnDescriptor
...
+  }
+
+CONVERT_PHYSICAL_TYPE_COPY_PTR(bool, uint8_t)
+CONVERT_PHYSICAL_TYPE_COPY_DATA(int32_t, uint8_t)
+CONVERT_PHYSICAL_TYPE_COPY_DATA(int32_t, int8_t)
+CONVERT_PHYSICAL_TYPE_COPY_DATA(int32_t, uint16_t)
+CONVERT_PHYSICAL_TYPE_COPY_DATA(int32_t, int16_t)
+CONVERT_PHYSICAL_TYPE_COPY_PTR(int32_t, uint32_t)
+CONVERT_PHYSICAL_TYPE_COPY_PTR(int32_t, int32_t)
+CONVERT_PHYSICAL_TYPE_COPY_PTR(int64_t, uint64_t)
```
Comment (wesm) [inline cpp/src/arrow/parquet/reader.cc]: Is this right for the unsigned ints?

**Comment by wesm** [comment]
Didn't see any major issues w/ the code -- main thing we have to figure out is the Parquet 2.0 metadata problem -- if using the new ConvertedType values will cause issues, we may have to live with all integers coming back as either INT32 or INT64 on the read path -- we could add a setting that force integers to be casted to a primitive type. Either way the code you've got here for performing the conversion will be useful, there may just be some refactoring to do.

File: cpp/src/arrow/parquet/parquet-io-test.cc (hunk 1 above)
Comment (xhochy) [inline cpp/src/arrow/parquet/parquet-io-test.cc]: So this probably needs a Parquet 1/2 switch somewhere in the WriterProperties?

File: cpp/src/arrow/parquet/writer.cc (hunk 2 above)
Comment (xhochy) [inline cpp/src/arrow/parquet/writer.cc]: Probably nothing different will happen if we just copy and cast these integers. The ones in the range  0<=x<=INT_MAX will be same in both, otherwise the ones >INT_MAX will overflow and return as negative. If both sides support Parquet 2, you should be able to pass all possible uint32 values like this to the reader.

File: cpp/src/arrow/parquet/parquet-io-test.cc (hunk 1 above)
Comment (wesm) [inline cpp/src/arrow/parquet/parquet-io-test.cc]: Yeah, I believe so. The problem is, at the moment Parquet 2 is not "in production". Even parquet-mr is still writing Parquet 1 files by default, so we should also default to parquet 1.x behavior for the time being

File: cpp/src/arrow/parquet/writer.cc (hunk 2 above)
Comment (wesm) [inline cpp/src/arrow/parquet/writer.cc]: The Parquet format is not explicit on this matter, I suspect that it's expecting you to do a static cast to one of the signed types. Probably merits a mailing list thread to get some answers

File: cpp/src/arrow/parquet/reader.cc (hunk 3 above)
Comment (wesm) [inline cpp/src/arrow/parquet/reader.cc]: I understand now that this is effectively "encoding" unsigned ints in a signed integer. Per above we should find out what other reader implementations are expecting (cast versus reinterpreting the bytes)

**Comment by xhochy** [comment]
Waiting for https://github.com/apache/parquet-cpp/pull/119 to implement writer code whose behaviour depends on the Parquet version.

**Comment by xhochy** [comment]
Actually looking at the code again, this patch is working irrelevant for which Parquet version we are writing. The version handling has to be done in `cpp/src/arrow/parquet/schema.cc` as only the supported logical types are different between Parquet 2 and Parquet 1, the physical types are the same. Thus the write and read path are the same, only the metadata (and the tests depending on that) are changing.

File: cpp/src/arrow/parquet/writer.cc (hunk 2 above)
Comment (xhochy) [inline cpp/src/arrow/parquet/writer.cc]: There is no difference between the two cast behaviours here, see:

```
#include <cstdint>
#include <iostream>
#include <limits>

int main() {
    uint32_t array[3] = {0, std::numeric_limits<int32_t>::max() + 1, std::numeric_limits<uint32_t>::max()};
    std::cout << array[0] << " " << array[1] << " " << array[2] << std::endl;
    int32_t* signed_array = reinterpret_cast<int32_t*>(array);
    std::cout << signed_array[0] << " " << signed_array[1] << " " << signed_array[2] << std::endl;
    std::cout << static_cast<int32_t>(array[0]) << " " << static_cast<int32_t>(array[1]) << " " << static_cast<int32_t>(array[2]) << std::endl;
}
```

Output:

```
0 2147483648 4294967295
0 -2147483648 -1
0 -2147483648 -1
```

**Comment by xhochy** [comment]
@wesm Is the PR ok for you or do we wait on other input?

I would proceed in a separate PR to implement the metadata switch between Parquet 1 and Parquet 2 in `cpp/src/arrow/parquet/schema.cc`.

Also when this PR is merged, I will directly open a new one for String I/O (that one is done but depends on this and I do not want to add a huge chunk of code to a nearly finished PR).

**Comment by wesm** [comment]
I'll review again shortly -- either way Parquet 1 must be the default as Parquet 2 is not yet widespread (or even fully specified, as far as I know)

**Comment by xhochy** [comment]
Made https://github.com/apache/arrow/pull/93 which should go in first then to have the Parquet 1 vs 2 switch I would then use here to test the different type roundtrips depending on the Parquet version.

**Comment by wesm** [comment]
Sorry about the delay, I will look to merge this tomorrow morning

File: cpp/src/arrow/parquet/writer.cc (hunk 2 above)
Comment (wesm) [inline cpp/src/arrow/parquet/writer.cc]: I guess I'm still thinking about Parquet 1.0. If you store the bytes in unsigned integers without a static cast, then Parquet 1.0 readers won't be able to recover the effective values:

So in your example

```
In [11]: i4_max = np.iinfo('i4').max

In [12]: u4_max = np.iinfo('u4').max

In [13]: arr = np.array([0, i4_max + 1, u4_max], dtype=np.uint32)

In [14]: arr
Out[14]: array([         0, 2147483648, 4294967295], dtype=uint32)

In [15]: arr.view(np.int32)
Out[15]: array([          0, -2147483648,          -1], dtype=int32)
```

2 out of the 3 values are lost because Parquet 1.0 readers don't know about unsigned integers.

Meanwhile, you can embed uint32 in int64 and get the correct data out:

```
In [16]: arr.astype(np.int64)
Out[16]: array([         0, 2147483648, 4294967295])
```

the compromise then is that uint32 must be stored in int64, and the other smaller unsigned integers must be stored in int32. uint64 values exceeding INT64_MAX can't be stored. In SQL databases this is the BIGINT type and so we can't handle large uint64 ints anyway.

So in summary, unlike the metadata we need to have divergent behavior around unsigned integers with Parquet 1.0 (i.e. the default right now) and 2.0

Comment (xhochy) [inline cpp/src/arrow/parquet/writer.cc]: Ah, that makes sense. Then I'll add a write path for Parquet 1.0 that saves UINT32 columns as INT64 and for Parquet 2.0, we store them as INT32 with UINT32 as logical type.

**Comment by xhochy** [comment]
Also pushed now the string support as this would otherwise involve quite some git magic to keep them in sync.

**Comment by wesm** [comment]
Uwe -- would you mind pushing this review to gerrit? You can adapt the Kudu instructions

http://kudu.apache.org/docs/contributing.html#_initial_setup_for_gerrit

(remember to replace instances of "kudu" with "arrow" in the commands)

You will need to squash your commits into a single patch. Thanks

**Comment by xhochy** [comment]
@wesm done: https://gerrit.cloudera.org/#/c/3481/

**Comment by wesm** [comment]
Thanks. If you trigger a build then things should work now. I'm working on the review, should finish tomorrow hopefully

**Comment by xhochy** [comment]
Yep, Travis is finally happy.

**Comment by xhochy** [comment]
Depends on https://github.com/apache/parquet-cpp/pull/128

**Comment by xhochy** [comment]
Travis still fails as the travis job for parquet-cpp/OSX had a hiccup and we only have the newest parquet-cpp version for Linux.

**Comment by wesm** [comment]
+1 -- we conducted the code review on gerrit https://gerrit.cloudera.org/#/c/3481.
